Plutarch/FFI.hs

```diff
@@ -0,0 +1,145 @@
+import qualified GHC.TypeLits as TypeLits
+import qualified Generics.SOP as SOP
+import Plutarch.Bool (PBool)
+import Plutarch.Builtin (PAsData, PData)
+import Plutarch.ByteString (PByteString)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal (
+  ClosedTerm,
+  PDelayed,
+  PType,
+  RawTerm (RCompiled),
+  Term (Term),
+  TermResult (TermResult),
+  asClosedRawTerm,
+  compile',
+  (:-->),
+ )
+import Plutarch.Internal.PlutusType (PlutusType (PInner))
+import Plutarch.String (PString)
+import Plutarch.Unit (PUnit)
+import Plutus.V1.Ledger.Scripts (Script (unScript), fromCompiledCode)
+import PlutusTx.Builtins.Internal (BuiltinBool, BuiltinByteString, BuiltinData, BuiltinUnit)
+import PlutusTx.Code (CompiledCode, CompiledCodeIn (DeserializedCode))
+import PlutusTx.Prelude (BuiltinString)
+import UntypedPlutusCore (fakeNameDeBruijn)
+import qualified UntypedPlutusCore as UPLC
+
+data ForallPhantom :: Type
+data PhorallPhantom :: PType
+
+data Delayed :: Type -> Type
+
+-- | Compile and export a Plutarch term so it can be used by `PlutusTx.applyCode`.
+foreignExport :: p >~< t => ClosedTerm p -> CompiledCode t
+foreignExport = unsafeForeignExport
+
+-- | Import compiled UPLC code (such as a spliced `PlutusTx.compile` result) as a Plutarch term.
+foreignImport :: p >~< t => CompiledCode t -> ClosedTerm p
+foreignImport = unsafeForeignImport
+
+-- | Seriously unsafe, may fail at run time or result in unexpected behaviour in your on-chain validator.
+unsafeForeignExport :: ClosedTerm p -> CompiledCode t
+unsafeForeignExport t = DeserializedCode program Nothing mempty
+  where
+    program =
+      UPLC.Program () (UPLC.Version () 1 0 0) $
+        UPLC.termMapNames fakeNameDeBruijn $
+          compile' $
+            asClosedRawTerm t
+
+-- | Seriously unsafe, may fail at run time or result in unexpected behaviour in your on-chain validator.
+unsafeForeignImport :: CompiledCode t -> ClosedTerm p
+unsafeForeignImport c = Term $ const $ TermResult (RCompiled $ UPLC.toTerm $ unScript $ fromCompiledCode c) []
+
+-- | Equality of inner types - Plutarch on the left and Haskell on the right.
+type p >~< t = PlutarchInner p PhorallPhantom ~~ PlutusTxInner t ForallPhantom
+
+type family a ~~ b :: Constraint where
+  ForallPhantom ~~ _ = ()
+  _ ~~ ForallPhantom = ()
+  Delayed a ~~ Delayed b = a ~~ b
+  a ~~ b = a ~ b
+
+type family PlutarchInner (p :: PType) (any :: PType) :: Type where
+  PlutarchInner PBool _ = BuiltinBool
+  PlutarchInner PInteger _ = Integer
+  PlutarchInner PString _ = Text
+  PlutarchInner PByteString _ = ByteString
+  PlutarchInner PData _ = BuiltinData
+  PlutarchInner PUnit _ = ()
+  PlutarchInner PhorallPhantom _ = ForallPhantom
+  PlutarchInner (PAsData a :--> PAsData b) x = PlutarchInner (PData :--> PData) x
+  PlutarchInner (PAsData a :--> b) x = PlutarchInner (PData :--> b) x
+  PlutarchInner (a :--> b) x = PlutarchInner a x -> PlutarchInner b x
+  PlutarchInner (PDelayed a) x = Delayed (PlutarchInner a x)
```

Comment:
I'm honestly not sure why you support this given that the resulting term can't be used from `PlutusTx`, nor is the inverse ever possible.

Reply:
That's there to bridge a mismatch between the Scott encodings of product types. PlutusTx always delays them for some reason:
```
testCase "import a pair" $
            printEvaluatedTerm
              ( foreignImport (PlutusTx.liftCode ("foo" :: BuiltinString, 4 :: Integer)) ::
                  Term _ (PDelayed (PPair PString PInteger))
              )
              @?= Right "(program 1.0.0 (delay (\\i0 -> i1 \"foo\" 4)))"
```